Post Featured Image: Fix borders after addition of overlay feature

aaronrobertshaw opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/44262

Related:
- https://github.com/WordPress/gutenberg/pull/44276

## What?

Fixes:
- Application of theme.json or global styles supplied borders to the new overlay element
- Application of border classes and styles for the overlay when configured at individual block level
- Disjoint between the markup rendered for the overlay in the editor vs the frontend

## Why?

Allows border radius to be applied to Post Featured Image blocks using overlays.

## How?

- Adds the block's overlay classname to the border block supports feature-level selectors meaning the theme.json/global styles target that as well as the inner image.
- Updates the generation of the overlay markup to apply border classes and styles
- Applies same logic as server-side render callback to only render overlay in the editor when appropriate
- Adds border classes to the overlay in the editor the same as it gets on the frontend
- Added a couple of fallback styles via `:where` to provide solid border style as that isn't officially supported on Post Featured Image blocks.

## Testing Instructions
1. Before checking out this PR, open the editor, add a Post Featured image block, apply an overlay to it and save
2. Update your theme.json to style Post Featured Image blocks with a border radius or set one via Global Styles > Blocks > Post Featured Image > Layout > Border radius.
3. View your post on the front end. You should see that the overlay element does not have a border radius applied to it.
4. Back in the editor attempt to provide custom borders on the individual block. Note these won't be applied correctly to the overlay on the frontend.
5. Clear the block level border configuration and save.
6. Check out this PR, reload the frontend and the overlay should now have the border styles in your theme.json or global styles.
7. In the editor attempt to configuration custom borders again for the individual block and ensure they are applied correctly.

Example theme.json snippet:
```json
			"core/post-featured-image": {
				"border": {
					"radius": {
						"topLeft": "5rem",
						"bottomRight": "5rem"
					},
					"color": "cyan",
					"width": "1.5em"
				}
			},
```


## Screenshots or screencast <!-- if applicable -->


https://user-images.githubusercontent.com/60436221/191416084-17711b47-d14e-47c0-a9e9-ac4e2fad7f3e.mp4

aaronrobertshaw committed:
Add overlay to post featured image border selectors

mikachan reviewed (approved):
Thanks for this! Works well for me 🎉

aaronrobertshaw committed:
Fix borders on overlays for individual blocks

aaronrobertshaw committed:
Provide a fallback border style

aaronrobertshaw committed:
Make the editor markup match frontend

Overlay was being rendered in editor regardless of if there should be one and it didn't have border classes applied to the overlay.

aaronrobertshaw commented:
I've increased the scope of this PR from fixing only the theme.json/global styles handling of borders for the Post Featured Image block's overlay to also fixing the individual block classes and styles, ensuring the editor markup is as close as possible to the frontend, and adding some default border style fallbacks.

It could do with some extra testing if you have the time @mikachan 🙏

RahiDroid commented on packages/block-library/src/post-featured-image/style.scss:
```diff
@@ -12,6 +12,11 @@
+	:where(img),
+	:where(.wp-block-post-featured-image__overlay) {
+		border-style: solid;
+	}
```
Elements seem to inherit the border-style fallback from https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/common.scss#L134-L136. Do we still need to specify this here?

aaronrobertshaw replied:
The fallback style you linked will only be applied when an inline style is set. The intent of these styles was to apply a fallback for when only border color or width were set via theme.json 

Our image-related blocks were requested to only support border color and width which means there's no control to explicitly set style at the individual or global styles levels. We might need to revisit that decision in a follow-up to this.

RahiDroid replied:
Makes sense. Thank you for the explanation! 👍

RahiDroid commented on packages/block-library/src/post-featured-image/overlay.js:
```diff
@@ -47,20 +47,23 @@ const Overlay = ( {
-				) }
-				style={ overlayStyles }
-			/>
+			{ !! dimRatio && (
```
Nice catch!

mikachan commented on packages/block-library/src/post-featured-image/style.scss:
```diff
@@ -12,6 +12,11 @@
 
+	:where(img),
+	:where(.wp-block-post-featured-image__overlay) {
+		border-style: solid;
```
Could you explain why we need to apply a solid border style by default? With this change, I'm seeing a solid border on every featured image:

![image](https://user-images.githubusercontent.com/1645628/191490255-f2936967-9665-4411-86dd-a62c3cc3c053.png)

![image](https://user-images.githubusercontent.com/1645628/191490480-86225930-2820-44dd-8380-4681a41c22d0.png)

![image](https://user-images.githubusercontent.com/1645628/191490751-406b4e1a-abc2-414a-ab1f-4d4bbf964a6b.png)

aaronrobertshaw replied:
Apologies, I missed this as the image I was testing with had black edges. 

It was added because if you only set a border color and width via theme.json then no border style is set as core adds `img { border: none }`. Setting the border style via theme.json then introduces other edge cases when the block doesn't support border style, as is the case with our image-related blocks at the moment. 

I've removed (https://github.com/WordPress/gutenberg/pull/44286/commits/b0eeade929546472aabcbecbf4ba38466658b0bd) these fallbacks and the decision to omit border style support from the post featured image block can be revisited in a follow-up.

mikachan replied:
Ahh right, thanks for explaining. This makes sense and sounds like a good follow-up.

mikachan reviewed (commented):
This is looking good! I like how you're conditionally showing the overlay based on `dimRatio` in overlay.js.

I just have one inline question.

aaronrobertshaw committed:
Remove fallback for theme.json border style

This would need a border-width: 0 as well to avoid the border it adds. Doing so would mean multiple edits required at the Global Styles level worsening UX. Setting the style via theme.json comes with further edge cases as well.

aaronrobertshaw committed:
Reduce unnecessary specificity in styles

aaronrobertshaw commented on packages/block-library/src/post-featured-image/editor.scss:
```diff
@@ -6,87 +6,85 @@
 		backdrop-filter: none; // Removes background blur so the overlay's actual color is visible.
 	}
 
-	&.wp-block-post-featured-image {
```
While doing some further testing around this block's styles I noticed that were duplicating the block's class selector here. Removing it didn't appear to impact the block visually in either editor and inspecting via devtools still showed all the same styles applied.

As a result, I've cleaned these styles up. Including reducing the specificity on the [button](https://github.com/WordPress/gutenberg/pull/44286/files#diff-5d5782edc25f6b5bb715554f10a066b6433a93fead2a1b99bbe104a6eb5edd47L23) that also didn't need it.

mikachan replied:
This tests well for me, I can also see no difference in the block's styles in either editor 👍

andrewserong reviewed (approved):
This is testing nicely for me, too, and the code change looks good 👍. I noticed one issue that's unrelated to this change (also appears to be present on `trunk`). In the editor view running TwentyTwentyTwo, when I set a border at the individual block level, it appears that the `solid` rule isn't sticking in the editor, but is on the site frontend. The effect is that with a dimRatio set, it looks like the border is transparent in the editor, but solid on the frontend (or to be more specific, on the site frontend the width of the image does not expand to overlap with the border):

![image](https://user-images.githubusercontent.com/14988353/192407302-4aa085bf-0efe-446c-9ed2-172c12e43fbc.png)

Since that issue is also present on `trunk`, I don't think it's a blocker to landing this PR — we can always look into that separately.

Other than that, with this change applied, it looks like all the border settings are being applied correctly in global styles and at the individual block level as far as I can tell. LGTM!

aaronrobertshaw commented:
Thanks for the review and thorough testing as always @andrewserong, appreciate it 👍 

Unfortunately, I haven't been able to replicate the issue you reported. Either on trunk or this PR branch. I've tried not only the main TwentyTwentyTwo theme but also its variants. I can't seem to achieve the transparent border at all in the editor.

As suggested, I'll merge this PR to get the fixes in.  If you can provide further details on how I can replicate the problem, I'll happily fix it in a follow-up.


https://user-images.githubusercontent.com/60436221/192473864-73e75d5b-1822-40aa-bd72-62a29342c134.mp4

andrewserong commented:
Good idea merging this one in first, and thanks for testing the issue 👍

The closest I can seem to get for reproduction instructions is that I'm running WordPress 6.0.2 in my local environment, and I can see the following two rules being applied to the `img` element from core's `common.css` file:

<img width="419" alt="image" src="https://user-images.githubusercontent.com/14988353/192662068-4ae767f6-05c0-4624-844a-68f73a68d8f7.png">

Because they're using the `border:` shorthand, I believe they wind up overriding the `where()` rule for `border-style: solid`? If I disable these rules in the editor, then the transparent border problem goes away:

![image](https://user-images.githubusercontent.com/14988353/192662299-051db6dd-817a-451e-9c85-d56d837d5814.png)

The thing that has me stumped, though, is why this issue only appears to be happening in my environment, and not yours 🤔

aaronrobertshaw commented:
> The thing that has me stumped, though, is why this issue only appears to be happening in my environment, and not yours 

Thanks for retesting and confirming the issue @andrewserong 👍 

I'd forgotten that Global Styles were saved per theme so when flicking between themes trying to test this, I didn't realise I already had styles for the Post Featured Image block. After resetting Global Styles for TwentyTwentyTwo and all its variants, I could replicate the issue.

A fix is up in https://github.com/WordPress/gutenberg/pull/44520

andrewserong commented:
Excellent, thanks for the quick follow-up! 🙇

michalczaplinski commented:
Cherry-picked for the 6.1 release in https://github.com/WordPress/gutenberg/pull/44656
